### PR TITLE
feat(md-enhance): support playground

### PR DESCRIPTION
**Pull request by Mister-Hope**

### CODE REVIEW

File: packages/md-enhance/src/node/markdown-it/playground.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import type { default as Token } from "markdown-it/lib/token";
+import type { PlaygroundFiles } from "../../shared/playground";
+import { IMPORT_MAP_KEY } from "../../shared/playground";
+
+const extensions = ["html", "js", "ts", "vue", "jsx", "tsx", "json"];
+
+// export const playground: PluginSimple = (md) => {
+export const playground =
+  (defaultImportMap?: string): PluginSimple =>
+  (md) =>
```
Comment (Mister-Hope) [inline packages/md-enhance/src/node/markdown-it/playground.ts:15]: This should be plugin with options.

File: packages/md-enhance/src/node/markdown-it/playground.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
+import { hash } from "@vuepress/utils";
```
Comment (Mister-Hope) [inline packages/md-enhance/src/node/markdown-it/playground.ts:1]: You may rewrite this plugin by editing currect tab plugin, we may have `@file` `@setting` and other sperator instead playground container.

File: packages/md-enhance/src/shared/options.ts [hunk 3]
```diff
@@ -232,13 +233,22 @@ export interface MarkdownEnhanceOptions {
    */
   presentation?: PresentationOptions | boolean;
 
-  /**
+  /*
```
Comment (Mister-Hope) [inline packages/md-enhance/src/shared/options.ts:236]: Do not change this, comment starting with `/**` has special meanings

File: packages/md-enhance/src/client/utils/playground/assign.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      !Array.isArray(originObject[property])
+    )
+      deepAssignReverse(
+        originObject[property] as IAnyObject,
+        assignObject[property] as IAnyObject
+      );
+  });
+
+  return deepAssignReverse(...assignObjects, assignObject);
+};
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/utils/playground/assign.ts:77]: I am not sure, do deepAssign only provided at node side? If so you move that to shared, do not have dumplicated functions in a single repo.

File: packages/md-enhance/src/client/utils/playground/utils.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
+// import { deepAssign } from "vuepress-shared";
+import { deepAssign } from "./assign";
+
+import {
+  PlaygroundOptions,
+  PlaygroundFiles,
+  IMPORT_MAP_KEY,
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/utils/playground/utils.ts:7]: We prefer to use `import type`

File: packages/md-enhance/src/client/utils/playground/utils.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  return files;
+}
+
+export function parsePlaygroundSettings(settings: string): PlaygroundOptions {
+  const settingsObj = JSON.parse(
+    decodeURIComponent(settings)
+  ) as PlaygroundOptions;
+
+  return deepAssign({}, APP_PLAYGROUND_SETTINGS, settingsObj);
+}
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/utils/playground/utils.ts:62]: I am definitely sure I see these links in `@vue/repl`.

If you are copying some files from a package, you should add comments declaring it's original file link in file header, and you NEED to copy the MIT license here using comment to follow MIT license, that's important.

File: packages/md-enhance/src/client/styles/playground.scss [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+      }
+
+      // stylelint-disable max-nesting-depth
+      &:hover {
+        .icon {
+          transition: transform 0.15s cubic-bezier(0.2, 0, 0.13, 2);
+          transform: scale(1.2);
+        }
+      }
+      // stylelint-enable max-nesting-depth
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/styles/playground.scss:43]: This is no longer needed

File: packages/md-enhance/src/client/styles/playground.scss [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+      display: inline-flex;
+      margin-left: 10px;
+
+      .icon {
+        width: 1.5rem;
+        height: 1.5rem;
+        fill: #777;
+      }
+
+      // stylelint-disable max-nesting-depth
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/styles/playground.scss:36]: This is no longer needed

File: packages/md-enhance/src/client/styles/playground.scss [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+
+  .title-container {
+    display: flex;
+    flex-flow: row wrap;
+
+    padding: 8px 10px;
+    border-bottom: 1px solid;
+    border-bottom-color: var(--code-border-color);
+
+    background-color: var(--code-demo-header-bg-color, #eee);
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/styles/playground.scss:57]: As people may only enable playground, but not code demo, you may declare a new css variable here

File: packages/md-enhance/src/client/styles/playground.scss [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+      text-overflow: ellipsis;
+      white-space: nowrap;
+    }
+  }
+
+  .preview-container {
+    position: relative;
+    overflow: hidden;
+    height: 300px;
+    background-color: var(--bg-color);
```
[the remark this answers is not in the document]
Comment (Mister-Hope) [reply packages/md-enhance/src/client/styles/playground.scss:76]: If you found I missed some styles in other places, PR welcomed.

File: packages/md-enhance/src/client/components/playground/Playground.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+import { IMPORT_MAP_KEY } from "../../../shared/playground";
+
+export default defineComponent({
+  // eslint-disable-next-line vue/multi-word-component-names
+  name: "Playground",
+
+  components: {
+    PlayFile,
+    PlayImports,
+    PlaySettings,
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/playground/Playground.ts:22]: So what are these empty components for?

File: packages/md-enhance/src/client/components/icons.ts [hunk 12]
```diff
@@ -63,4 +63,8 @@ export const CODEPEN_SVG =
...
+
+export const CODE_SVG = `<svg t="1651043677975" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="11716" width="200" height="200"><path d="M240.496 272c-13.504 0-25.664 5.632-34.384 14.608l-0.048-0.048L16.848 475.76A47.664 47.664 0 0 0 0 512c0 13.504 5.616 25.664 14.592 34.368l-0.032 0.064 192 192 0.048-0.064A47.68 47.68 0 0 0 240 752a48 48 0 0 0 48-48c0-12.992-5.216-24.752-13.616-33.392l0.048-0.048-158.304-158.32 157.808-157.808-0.048-0.048A47.808 47.808 0 0 0 288.496 320a48 48 0 0 0-48-48z m784 240c0-14.56-6.608-27.44-16.848-36.24L818.432 286.56l-0.032 0.048A47.872 47.872 0 0 0 784 272a48 48 0 0 0-48 48c0 13.504 5.632 25.664 14.608 34.384l-0.048 0.048 157.808 157.808-158.32 158.32 0.048 0.048A47.808 47.808 0 0 0 736.48 704a48 48 0 0 0 48 48c12.992 0 24.752-5.216 33.408-13.632l0.048 0.064 192-192-0.048-0.064A47.68 47.68 0 0 0 1024.496 512zM640 128c-20.8 0-38.496 13.232-45.168 31.712L339.2 830.784a48 48 0 0 0 89.968 33.504L684.8 193.216A48 48 0 0 0 640 128z" fill="" p-id="11717"></path></svg>`;
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/icons.ts:70]: You should use `svgo` to compress the SVG string before adding it here.

File: packages/md-enhance/src/shared/playground.ts [hunk 13]
```diff
@@ -0,0 +1,109 @@
+import type { SFCOptions } from "@vue/repl";
```
Comment (Mister-Hope) [inline packages/md-enhance/src/shared/playground.ts:1]: Chinese js docs should be added in this file.

File: packages/md-enhance/src/client/components/playground/Playground.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+import { parsePlaygroundSettings } from "../../utils/playground";
+import { IMPORT_MAP_KEY } from "../../../shared/playground";
+
+export default defineComponent({
+  // eslint-disable-next-line vue/multi-word-component-names
+  name: "Playground",
+
+  components: {
+    PlayFile,
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/playground/Playground.ts:20]: If you are using render function you do not need to declare components here, so you should remove it.

Also I suggested names like ExternalPlayground and PlaygroundFile

File: packages/md-enhance/src/client/components/playground/Playground.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    PlayImports,
+    PlaySettings,
+    PlaygroundExternal,
+    PlaygroundInternal,
+  },
+
+  props: {
+    id: { type: String, required: true },
+    title: { type: String, default: "" },
+    config: { type: String, default: null },
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/playground/Playground.ts:30]: Is this legal?

File: packages/md-enhance/src/client/components/playground/Playground.ts [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+  setup(props) {
+    const settings = parsePlaygroundSettings(props.settings);
+    const mode = settings.mode;
+
+    const encodedKey = encodeURIComponent(IMPORT_MAP_KEY);
+
+    return (): (VNode | null)[] => [
+      h(ClientOnly, null, [
+        mode === "internal"
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/playground/Playground.ts:42]: You may change the first option in h, and avoid writing it twice here

File: packages/md-enhance/src/client/components/playground/PlaygroundExternal.ts [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    config: { type: String, required: true },
+    settings: { type: String, default: "{}" },
+    id: { type: String, required: true },
+  },
+
+  setup(props) {
+    const playgroundContainer = ref<HTMLElement | null>(null);
+    const iframe = ref<HTMLElement | null>(null);
+
+    const loading = ref<boolean>(true);
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/playground/PlaygroundExternal.ts:24]: The Boolean is unnecessary here

File: packages/md-enhance/src/client/components/playground/PlaygroundExternal.ts [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+    const playgroundOptions = parsePlaygroundSettings(props.settings);
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/playground/PlaygroundExternal.ts:26]: A reminder This is not reactive. So why the following variables is wrapped with computed?

File: packages/md-enhance/src/client/components/playground/PlaygroundExternal.ts [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+    const previewLink: Ref<string> = computed(() => {
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/playground/PlaygroundExternal.ts:28]: The ref string is not necessary either

File: packages/md-enhance/src/client/components/playground/PlaygroundExternal.ts [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+      const { link } = usePlaygroundExternal(
+        props.config,
+        playgroundOptions.external || {}
+      );
+
+      return link;
+    });
+
+    const hideLoading = () => {
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/playground/PlaygroundExternal.ts:37]: If a function is kind of simple like this, and we are not using it in multiple places, we may prefer to write the function directly in render function

File: packages/md-enhance/src/client/components/playground/PlaygroundInternal.ts [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+import "../../styles/playground.scss";
+import { LOADING_SVG, CODE_SVG } from "../icons";
+import { usePlaygroundExternal } from "../../composables/playground";
+import { parsePlaygroundSettings } from "../../utils/playground";
+
+export default defineComponent({
+  name: "PlaygroundInternal",
+
+  props: {
+    title: { type: String, default: "" },
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/playground/PlaygroundInternal.ts:17]: Vue style guide suggests declaring required props first

File: packages/md-enhance/src/client/components/playground/PlaygroundInternal.ts [hunk 22]
```diff
@@ -0,0 +1,132 @@
+import { defineComponent, h, onMounted, ref, reactive } from "vue";
+import type { VNode } from "vue";
+
+import type { ReplProps } from "@vue/repl";
+import { ReplStore, Repl } from "@vue/repl";
```
Comment (Mister-Hope) [inline packages/md-enhance/src/client/components/playground/PlaygroundInternal.ts:5]: I think this component may might be large, so you may dynamic import this or the whole component. Vue3 provides a defineAsyncComponent

File: packages/md-enhance/src/shared/options.ts (hunk 3 above)
Comment (FuckDoctors) [reply packages/md-enhance/src/shared/options.ts:236]: I missed 1 `*` when merging...

**Pull request merged**